### ISSUE TITLE
Investigate the idea of `App.query( Type )` providing a hint as to its return type

### ISSUE THREAD

**Issue by davep** [Textualize/textual#867]
As per the in-person conversation with Will: it's currently possible to use `App.query` where an actual type is passed as the query. For example:

```python
buttons = self.query( Button )
```

as opposed to, say:

```python
buttons = self.query( "Button" )
```

The question here is, in the former case, is it then possible to type-hint the code such that a type checking would know that a `DOMQuery` of the given type is being returned.

By way of illustration:

```python
    @property
    def my_buttons(self) -> DOMQuery[Button]:
        """The buttons in my app.

        :type: DOMQuery[Button]
        """
        return self.query( Button )
```

would have any type checker happy that `self.query( Button )` is returning `DOMQuery[Button]`.

Caveat: this came about after seeing some warnings from `pyright`, but not from `mypy`; so it's possible that this is simply chasing a `pyright` issue.

**Comment by davep** [comment]
Having dived into this somewhat, I think I'm happy with the idea that the motivation for this question has come from some odd working of `pyright` and my own editor setup. To illustrate, this is how my own setup under Emacs (which shows a type warning) sees `App.query`:

<img width="495" alt="Screenshot 2022-10-12 at 13 20 34" src="https://user-images.githubusercontent.com/28237/195341346-68f92928-fd9d-41db-b7b7-25b2f7876ce7.png">

vs in VSCode:

<img width="509" alt="Screenshot 2022-10-12 at 13 20 50" src="https://user-images.githubusercontent.com/28237/195341430-fb2c13e2-6029-4f73-937a-6dcd64b2b7b8.png">

I don't think there's really much need to pursue this any further.

**Comment by davep** [comment]
Bumping this back open as I've now been able to recreate this, with the same type interpretation as shown above in Emacs, if I use `pylance` under VSCode and have the type checking set to "strict". Also, if I set type checking for `pyright` in Emacs to "basic" the errors I'm seeing go away.

This could still just be an oddity with `pyright`, of course, but it would be good to be sure and/or add extra hints to Textual to ensure it's not a problem.
